dex/networks/eth/txdata.go

```diff
@@ -10,40 +10,134 @@ import (
+func ParseRefundData(calldata []byte, contractVersion uint32) ([32]byte, error) {
+	txDataHandler, ok := txDataHandlers[contractVersion]
+	if !ok {
+		return [32]byte{}, fmt.Errorf("contract version %v does not exist", contractVersion)
+	}
+
+	return txDataHandler.parseRefundData(calldata)
+}
 
-// ParseInitiateData accepts call data from a transaction that pays to a
-// contract with extra data. It will error if the call data does not call
-// initiate with expected argument types. It returns the array of initiations
-// with which initiate was called.
-func ParseInitiateData(calldata []byte) ([]swapv0.ETHSwapInitiation, error) {
+// PackInitiateData converts a list of Initiation to the call data for the
+// initiate function for the contract specified by contractVersion.
+func PackInitiateData(initiations []*Initiation, contractVersion uint32) ([]byte, error) {
+	txDataHandler, ok := txDataHandlers[contractVersion]
+	if !ok {
+		return nil, fmt.Errorf("contract version %v does not exist", contractVersion)
+	}
+
+	return txDataHandler.packInitiateData(initiations)
+}
+
+// PackRedeemData converts a list of Redemption to the call data for the redeem
+// function for the contract specified by contractVersion.
+func PackRedeemData(redemptions []*Redemption, contractVersion uint32) ([]byte, error) {
+	txDataHandler, ok := txDataHandlers[contractVersion]
+	if !ok {
+		return nil, fmt.Errorf("contract version %v does not exist", contractVersion)
+	}
+
+	return txDataHandler.packRedeemData(redemptions)
+}
+
+// PackRefundData converts a secret hash to the call data for the refund function
+// for the contract specified by contractVersion.
+func PackRefundData(secretHash [32]byte, contractVersion uint32) ([]byte, error) {
+	txDataHandler, ok := txDataHandlers[contractVersion]
+	if !ok {
+		return nil, fmt.Errorf("contract version %v does not exist", contractVersion)
+	}
+
+	return txDataHandler.packRefundData(secretHash)
+}
+
+type txDataHandler interface {
+	parseInitiateData([]byte) ([]*Initiation, error)
+	parseRedeemData([]byte) ([]*Redemption, error)
+	parseRefundData([]byte) ([32]byte, error)
+	packInitiateData([]*Initiation) ([]byte, error)
+	packRedeemData([]*Redemption) ([]byte, error)
+	packRefundData([32]byte) ([]byte, error)
+}
+
+var txDataHandlers = map[uint32]txDataHandler{
+	0: newTxDataV0(),
+}
+
+type txDataHandlerV0 struct {
+	initiateFuncName string
+	redeemFuncName   string
+	refundFuncName   string
+	parsedAbi        abi.ABI
+}
+
+func newTxDataV0() *txDataHandlerV0 {
+	parsedAbi, err := abi.JSON(strings.NewReader(swapv0.ETHSwapABI))
+	if err != nil {
+		panic(fmt.Sprintf("failed to parse abi: %v", err))
+	}
+
+	return &txDataHandlerV0{
+		initiateFuncName: "initiate",
+		redeemFuncName:   "redeem",
+		refundFuncName:   "refund",
+		parsedAbi:        parsedAbi,
+	}
+}
+
+func (t *txDataHandlerV0) parseInitiateData(calldata []byte) ([]*Initiation, error) {
 	decoded, err := parseCallData(calldata, swapv0.ETHSwapABI)
```

Comment:
This is actually the part that's gotta be refactored. The ABI need to be initialized in a package-level `init` function, and then accessed via map lookup throughout, including the e.g. `newV0contractor`.

Reply:
Ah, I neglected this part too.  `parseCallData` can take an `*abi.ABI` instead of a string and `&t.parsedAbi` can be provided to it.  I was ambivalent about package level vs txDataHandlerVX constructor though.

Looks like making this change will also lead to `verifySelector`, but that appear unused.

Reply:
I wanted to not touch that file since it was copied verbatim from the geth repo, but I guess there's no harm.